react-backend/client/src/cb-admin/pages/VisitsData.js

```diff
@@ -0,0 +1,338 @@
+  id: 'Visitor ID',
+  name: 'Name',
+  gender: 'Gender',
+  yob: 'Year of birth',
+};
+
+const colToState = invertObj(keyMap);
+
+const columns = Object.values(keyMap).filter(Boolean);
+
+
+export default class VisitsDataPage extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      users: [],
+      activities: [],
+      filters: {},
+      orderBy: '',
+      genderNumbers: { datasets: [], labels: [] },
+      visits: { datasets: [], labels: [] },
+      visitNumbers: { datasets: [], labels: [] },
+      ageGroups: { datasets: [], labels: [] },
+      activitiesGroups: { datasets: [], labels: [] },
+      errors: {},
+    };
+  }
+
+  componentDidMount() {
+    const pVisitors = Visitors.get(this.props.auth);
+    const pStats = Visitors.getStatistics(this.props.auth);
+
+    Promise.all([pVisitors, pStats])
+      .then(([resVisitors, resStats]) => {
+        this.props.updateAdminToken(resVisitors.headers.authorization);
+        const [visitsNumbers, genderNumbers, activitiesNumbers, ageGroups, activities] = resStats.data.result;
+
+        this.setState({
+          users: resVisitors.data.result,
+          visits: visitsNumbers,
+          visitNumbers: this.getVisitsWeek(visitsNumbers),
+          genderNumbers: this.getGendersForChart(genderNumbers),
+          activitiesGroups: this.getActivitiesForChart(activitiesNumbers),
+          ageGroups: this.getAgeGroupsForChart(ageGroups),
+          activities: activities.map(activity => activity.name),
+        });
+      })
+      .catch(error => this.setState({ errors: { general: 'Unknown error' } }));
+  }
+
+  onChange = e => this.setState({ [e.target.name]: e.target.value }, this.update)
+
+  getActivitiesForChart = (activities) => {
+    const activitiesData = {
+      labels: activities.map(el => el.name),
+      datasets: [
+        {
+          data: activities.map(el => el.count),
+          backgroundColor: repeat([
+            colors.highlight_primary,
+            colors.highlight_secondary,
+            colors.light,
+          ], activities.length),
+        },
+      ],
+    };
+    return activitiesData;
+  };
+
+  getGendersForChart = (genders) => {
+    const genderData = {
+      labels: genders.map(el => el.sex),
+      datasets: [
+        {
+          data: genders.map(el => el.count),
+          backgroundColor: [
+            colors.highlight_primary,
+            colors.highlight_secondary,
+          ],
+        },
+      ],
+    };
+    return genderData;
+  };
+
+  getVisitsWeek = (visits) => {
+    const now = moment();
+    const today = now.format('dddd');
+    const lastWeek = now.subtract(7, 'days');
+
+    const dN = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
+    const dayName = circShift(dN, -dN.indexOf(today));
+
+    const initObj = dayName.reduce((acc, d) => { acc[d] = 0; return acc; }, {});
+    const visitCount = visits
+      .map(v => v.date)
+      .slice().sort()
```

Comment:
What's the empty slice doing?

Reply:
cloning the array, because `sort` is mutative.